### PR TITLE
[MemProf] Prune unneeded non-cold contexts

### PR DESCRIPTION
**Pull request by teresajohnson**
We can take advantage of the fact that we subsequently only clone cold
allocation contexts, since not cold behavior is the default, and
significantly reduce the amount of metadata (and later ThinLTO summary
and MemProfContextDisambiguation graph nodes) by pruning unnecessary not
cold contexts when building metadata from the trie.

Specifically, we only need to keep notcold contexts that overlap the
longest with cold allocations, to know how deeply to clone those
contexts to expose the cold allocation behavior.

For a large target this reduced ThinLTO bitcode object sizes by about
35%. It reduced the ThinLTO indexing time by about half and the peak
ThinLTO indexing memory by about 20%.

### CODE REVIEW

File: llvm/lib/Analysis/MemoryProfileInfo.cpp [hunk 1]
```diff
@@ -51,6 +51,13 @@ cl::opt<bool> MemProfReportHintedSizes(
     "memprof-report-hinted-sizes", cl::init(false), cl::Hidden,
     cl::desc("Report total allocation sizes of hinted allocations"));
 
+// This is useful if we have enabled reporting of hinted sizes, and want to get
+// information from the indexing step for all contexts (especially for testing),
+// or have specified a value less than 100% for -memprof-cloning-cold-threshold.
+cl::opt<bool> MemProfKeepAllNotColdContexts(
+    "memprof-keep-all-not-cold-contexts", cl::init(false), cl::Hidden,
+    cl::desc("Disable pruning of non-cold contexts unneeded for cold cloning"));
```
Comment (snehasish) [inline llvm/lib/Analysis/MemoryProfileInfo.cpp:59]: nit: The current description is a bit hard to parse how about "Keep all non-cold contexts (increases cloning overheads)" instead of the current description. 

Also should the default be `true` (existing behaviour) since internal users set `-memprof-cloning-cold-threshold` to less than 100?

File: llvm/lib/Analysis/MemoryProfileInfo.cpp [hunk 2]
```diff
@@ -243,14 +256,35 @@ void CallStackTrie::convertHotToNotCold(CallStackTrieNode *Node) {
...
+    // 1) It is cold.
+    // 2) The immediate callee is the deepest point where we have an ambiguous
+    //    allocation type (i.e. the other callers that are cold need to know
+    //    that we have a not cold context overlapping to this point so that we
+    //    know how deep to clone).
+    // 3) MemProfKeepAllNotColdContexts is enabled, which is useful if we are
+    //    reporting hinted sizes, and want to get information from the indexing
+    //    step for all contexts, or have specified a value less than 100% for
+    //    -memprof-cloning-cold-threshold.
+    if ((AllocationType)Node->AllocTypes == AllocationType::Cold ||
```
Comment (snehasish) [inline llvm/lib/Analysis/MemoryProfileInfo.cpp:276]: Can this be rewritten as `Node->hasAllocType(AllocationType::Cold)` to use the new helpers?

Same for the usage below.

**Review by snehasish — COMMENTED** [review]
Still looking through the test.

File: llvm/lib/Analysis/MemoryProfileInfo.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp:59]: I will change the description. But I wanted to point out that we are not currently using -memprof-cloning-cold-threshold, only -memprof-matching-cold-threshold which doesn't rely on the metadata. I would like to keep the default here true to reduce overhead by default.

File: llvm/lib/Analysis/MemoryProfileInfo.cpp (hunk 2 above)
Comment (teresajohnson) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp:276]: ack

File: llvm/lib/Analysis/MemoryProfileInfo.cpp [hunk 3]
```diff
@@ -337,10 +371,16 @@ bool CallStackTrie::buildAndAttachMIBMetadata(CallBase *CI) {
...
-  // has more than one caller. Here the node being passed in is the alloc
-  // and it has no callees. So it's false.
-  if (buildMIBNodes(Alloc, Ctx, MIBCallStack, MIBNodes, false)) {
+  // The CalleeHasAmbiguousCallerContext flag is meant to say whether the
+  // callee of the given node has more than one caller. Here the node being
+  // passed in is the alloc and it has no callees. So it's false.
+  // Similarly, the last parameter is meant to say whether the callee of the
+  // given node is the deepest point where we have ambiguous alloc types, which
+  // is also false as the alloc has no callees.
+  bool DeepestAmbiguousAllocType = true;
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp:380]: ack

File: llvm/lib/Analysis/MemoryProfileInfo.cpp (hunk 1 above)
Comment (snehasish) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp:59]: Thanks for clarifying we only use the matching option for now.

Comment (teresajohnson) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp]: Updated the flag description per suggestion

File: llvm/lib/Analysis/MemoryProfileInfo.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp:380]: I started to do this but it made the uses more complicated inside buildMIBNodes and I think it ends up more complicated and hard to reason about overall. So I'd prefer to leave as-is.

File: llvm/lib/Analysis/MemoryProfileInfo.cpp (hunk 2 above)
Comment (teresajohnson) [reply llvm/lib/Analysis/MemoryProfileInfo.cpp]: Done here and below

File: llvm/unittests/Analysis/MemoryProfileInfoTest.cpp [hunk 4]
```diff
@@ -220,6 +221,48 @@ declare dso_local noalias noundef i8* @malloc(i64 noundef)
...
+  if (ExpectedVals.size() != arg->getNumOperands())
+    return PrintAndFail();
+
+  for (size_t I = 0; I < ExpectedVals.size(); I++) {
+    const auto &[ExpectedAllocType, ExpectedStackIds] = ExpectedVals[I];
+    MDNode *MIB = dyn_cast<MDNode>(arg->getOperand(I));
+    if (getMIBAllocType(MIB) != ExpectedAllocType)
+      return PrintAndFail();
+    MDNode *StackMD = getMIBStackNode(MIB);
+    EXPECT_NE(StackMD, nullptr);
```
Comment (snehasish) [inline llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:254]: This should be ASSERT since EXPECT doesn't stop the test and we would deref the nullptr on the next line.

File: llvm/unittests/Analysis/MemoryProfileInfoTest.cpp [hunk 5]
```diff
@@ -428,14 +469,65 @@ declare dso_local noalias noundef i8* @malloc(i64 noundef)
...
+
+  std::vector<std::pair<AllocationType, std::vector<unsigned>>> ExpectedVals = {
+      {AllocationType::Cold, {1, 2, 3, 4}},
+      {AllocationType::Cold, {1, 2, 3, 5, 6, 7}},
+      {AllocationType::NotCold, {1, 2, 3, 5, 6, 13}},
+      {AllocationType::Cold, {1, 2, 3, 8, 9, 10}},
+      {AllocationType::NotCold, {1, 2, 3, 8, 9, 14}}};
+
+  CallBase *Call = findCall(*Func, "call");
+  Trie.buildAndAttachMIBMetadata(Call);
```
Comment (snehasish) [inline llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:523]: Should we assert Call is not a nullptr before we pass it to the function?

File: llvm/unittests/Analysis/MemoryProfileInfoTest.cpp [hunk 6]
```diff
@@ -428,14 +469,65 @@ declare dso_local noalias noundef i8* @malloc(i64 noundef)
...
+  // cloned from:
+  Trie.addCallStack(AllocationType::NotCold, {1, 2, 3, 8, 9, 14});
+  // This NotCold context is not needed since the above is sufficient (we pick
+  // the first in sorted order).
+  Trie.addCallStack(AllocationType::NotCold, {1, 2, 3, 8, 9, 15});
+
+  // Neither of these two NotCold contexts are needed as the Cold contexts they
+  // overlap with are covered by longer overlapping NotCold contexts.
+  Trie.addCallStack(AllocationType::NotCold, {1, 2, 3, 12});
+  Trie.addCallStack(AllocationType::NotCold, {1, 2, 11});
```
Comment (snehasish) [inline llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:513]: Does it make sense to add one more test for the edge case like this -- `Trie.addCallStack(AllocationType::NotCold, {1, 16});`? I noticed that all of the interesting stuff starts from node 2 onwards.

**Review by snehasish — APPROVED** [review]
lgtm with some minor comments.

File: llvm/unittests/Analysis/MemoryProfileInfoTest.cpp (hunk 4 above)
Comment (teresajohnson) [reply llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:254]: I had ASSERT_NE here first but I get a compiler error:

llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:254:5: error: cannot initialize return object of type 'bool' with an rvalue of type 'void'
    ASSERT_NE(StackMD, nullptr);

File: llvm/unittests/Analysis/MemoryProfileInfoTest.cpp (hunk 6 above)
Comment (teresajohnson) [reply llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:513]: Done

File: llvm/unittests/Analysis/MemoryProfileInfoTest.cpp (hunk 5 above)
Comment (teresajohnson) [reply llvm/unittests/Analysis/MemoryProfileInfoTest.cpp:525]: Done
